src/main/java/org/fcrepo/storage/ocfl/DefaultOcflObjectSession.java

```diff
@@ -0,0 +1,477 @@
+ */
+public class DefaultOcflObjectSession implements OcflObjectSession {
+
+    private final String sessionId;
+    private final OcflRepository ocflRepo;
+    private final String ocflObjectId;
+    private final Path objectStaging;
+    private final ObjectMapper objectMapper;
+    private final Runnable deregisterHook;
+
+    private final OcflOption[] ocflOptions;
+    private final VersionInfo versionInfo;
+    private final Set<PathPair> deletePaths;
+
+    private String rootResourceId;
+    private boolean closed = false;
+    private boolean deleteObject = false;
+
+    public DefaultOcflObjectSession(final String sessionId,
+                                    final OcflRepository ocflRepo,
+                                    final String ocflObjectId,
+                                    final Path objectStaging,
+                                    final ObjectMapper objectMapper,
+                                    final Runnable deregisterHook) {
+        this.sessionId = sessionId;
+        this.ocflRepo = ocflRepo;
+        this.ocflObjectId = ocflObjectId;
+        this.objectStaging = objectStaging;
+        this.objectMapper = objectMapper;
+        this.deregisterHook = deregisterHook;
+
+        this.versionInfo = new VersionInfo();
+        this.deletePaths = new HashSet<>();
+        this.ocflOptions = new OcflOption[] {OcflOption.MOVE_SOURCE, OcflOption.OVERWRITE};
+
+        this.rootResourceId = loadRootResourceId();
+    }
+
+    @Override
+    public String sessionId() {
+        return sessionId;
+    }
+
+    @Override
+    public String ocflObjectId() {
+        return ocflObjectId;
+    }
+
+    @Override
+    public synchronized void writeResource(final String resourceId, final ResourceContent content) {
+        enforceOpen();
+
+        final var paths = resolvePersistencePaths(resourceId, content.getHeaders());
+
+        final var contentPath = encode(paths.getContentFilePath());
+        final var headerPath = encode(paths.getHeaderFilePath());
+
+        deletePaths.remove(contentPath);
+        deletePaths.remove(headerPath);
+
+        content.getHeaders().setContentPath(contentPath.path);
+
+        final var contentDst = createStagingPath(contentPath);
+        write(content.getContentStream(), contentDst);
+
+        final var headerDst = createStagingPath(headerPath);
+        writeHeaders(content.getHeaders(), headerDst);
+    }
+
+    @Override
+    public synchronized void deleteContentFile(final String resourceId, final ResourceHeaders headers) {
+        enforceOpen();
+
+        final var headerPath = encode(PersistencePaths.headerPath(rootResourceId(), resourceId));
+
+        if (newInSession(headerPath)) {
+            deleteResource(resourceId);
+        } else {
+            final var existingHeaders = readHeaders(resourceId);
+
+            deletePaths.add(encode(existingHeaders.getContentPath()));
+
+            final var headerDst = createStagingPath(headerPath);
```

Comment:
`deleteContentFile()` essentially supports the tombstone use case where you delete a content file and update the resource headers. The existing headers are loaded to locate the content file, and the argument headers are written because they are the updated headers that presumably indicate that the resource was deleted.

Reply:
Ah, I missed that it switched which headers object it was writing, sorry about that